core/templates/pages/diagnostic-test-player-page/diagnostic-test.model.ts

```diff
@@ -0,0 +1,256 @@
+
+      while (prerequisites.length > 0) {
+        let len = prerequisites.length;
+        let lastTopicId = prerequisites[len - 1];
+        prerequisites.splice(len - 1, 1);
+        if (ancestors.indexOf(lastTopicId) === -1) {
+          ancestors.push(lastTopicId);
+        }
+
+        prerequisites = prerequisites.concat(
+          this._topicIdToPrerequisiteTopicIds[lastTopicId]);
+      }
+      this._topicIdToAncestorTopicIds[topicId] = ancestors;
+    }
+  }
+
+  setTopicIdToSuccessorTopicIds(): void {
+    let topicIdToChildTopicId: TopicIdToRelatedTopicIds = {};
+
+    for (let topicId in this._topicIdToPrerequisiteTopicIds) {
+      topicIdToChildTopicId[topicId] = [];
+    }
+
+    for (let topicId in this._topicIdToPrerequisiteTopicIds) {
+      let prereq = this._topicIdToPrerequisiteTopicIds[topicId];
+      for (let prereqTopicId of prereq) {
+        topicIdToChildTopicId[prereqTopicId].push(topicId);
+      }
+    }
+
+    for (let topicId in topicIdToChildTopicId) {
+      let successors: string[] = [];
+      let children: string[] = topicIdToChildTopicId[topicId];
+
+      while (children.length > 0) {
+        let len = children.length;
+        let lastTopicId = children[len - 1];
+        children.splice(len - 1, 1);
+        if (successors.indexOf(lastTopicId) === -1) {
+          successors.push(lastTopicId);
+        }
+        children = children.concat(topicIdToChildTopicId[lastTopicId]);
+      }
+      this._topicIdToSuccessorTopicIds[topicId] = successors;
+    }
+  }
+
+  setCurrentTopicId(): void {
+    let topicIdToLengthOfRelatedTopicIds: {[topicId: string]: number} = {};
+    let lengthOfEligibleAncestorTopicIds: number;
+    let lengthOfEligibleSuccessorTopicIds: number;
+
+    for (let topicId in this._topicIdToPrerequisiteTopicIds) {
+      let ancestors = this._topicIdToAncestorTopicIds[topicId];
+      let successors = this._topicIdToSuccessorTopicIds[topicId];
+
+      lengthOfEligibleAncestorTopicIds = ancestors.filter((topic) => {
+        return (this._eligibleTopicIds.indexOf(topic) !== -1);
+      }).length;
+
+      lengthOfEligibleSuccessorTopicIds = successors.filter((topic) => {
+        return (this._eligibleTopicIds.indexOf(topic) !== -1);
+      }).length;
+
+      topicIdToLengthOfRelatedTopicIds[topicId] = Math.min(
+        lengthOfEligibleAncestorTopicIds, lengthOfEligibleSuccessorTopicIds);
+    }
+
+    let tempTopicId = Object.keys(topicIdToLengthOfRelatedTopicIds)[0];
+    let tempLenghtOfExpectedRemoval = (
+      topicIdToLengthOfRelatedTopicIds[tempTopicId]);
+
+    for (let topicId in topicIdToLengthOfRelatedTopicIds) {
+      if (
+        topicIdToLengthOfRelatedTopicIds[topicId] >
+          tempLenghtOfExpectedRemoval
+      ) {
+        tempLenghtOfExpectedRemoval = topicIdToLengthOfRelatedTopicIds[topicId];
+        tempTopicId = topicId;
+      }
+    }
+    this._currentTopicId = tempTopicId;
+  }
+
+  recordTopicPassed(): void {
```

Comment:
Again have a brief comment at the start of this function saying something like "If a topic is passed, that means we don't need to test any of the ancestor topics."

Reply:
Added